### PR TITLE
added structured logging specs, logger timestamp is utc

### PR DESCRIPTION
**Pull request by kreek**

### CODE REVIEW

**Comment by saneshark** [comment]
@kreek you might need to add rails to the gemspec to properly test the railtie stuff, just fyi.

I normally try to create gems agnostic of rails, but I think in this case it totally makes sense to go ahead and do it. I had to do something similar in va_rx gem with respect to ActiveModel.

File: lib/va/api/common/structured_logger/structured_logger.rb [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                { msg: head }.merge(tail.reduce({}, :update))
+              end
+          convert_quotes(h)
+        rescue StandardError => e
+          warn "failed to generate log message hash: #{e.message}"
+        end
+
+        def convert_quotes(h)
+          h.map do |k, v|
+            { k => v.is_a?(String) ? v.strip.tr('"', "'") : v }
```
Comment (saneshark) [inline lib/va/api/common/structured_logger/structured_logger.rb]: (nitpick) Since you have ActiveSupport, you could use Hash#transform_values here which might be slightly more readable.

File: lib/va/api/common/structured_logger/structured_logger.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+              else
+                head, *tail = args
+                { msg: head }.merge(tail.reduce({}, :update))
+              end
+          convert_quotes(h)
+        rescue StandardError => e
+          warn "failed to generate log message hash: #{e.message}"
+        end
+
+        def convert_quotes(h)
```
Comment (saneshark) [inline lib/va/api/common/structured_logger/structured_logger.rb]: (nitpick) should this and others be private methods?

File: lib/va/api/common/structured_logger/structured_logger.rb [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+        def error(*args, &block)
+          add(ERROR, message_hash(args, &block))
+        end
+
+        def fatal(*args, &block)
+          add(FATAL, message_hash(args, &block))
+        end
+
+        def message_hash(args, &_block)
```
Comment (saneshark) [inline lib/va/api/common/structured_logger/structured_logger.rb]: Do you think it would be more readable if each of these cases could be split out into methods that have more descriptive names as to what they are doing? I know you're just coercing arguments, and to me it makes sense to have a functional approach to doing this as you've done, but it could be hard to follow or understand. 

This is a nitpick though. And if you think a change could negatively affect performance, please ignore.

**Comment by saneshark** [comment]
@kreek I looked through this and only had a few minor nitpicks for the most part. I know in a lot of places you might just be overriding Logger methods, so for the most part, ignore most of my comments... The biggest question I have is are there things light coloring and highlighting that we should be able to easily toggle or disable? If we find that the app is hitting a bottleneck or something? I doubt this will be an issue, but maybe just something to think about.

File: lib/va/api/common/structured_logger/structured_logger.rb (hunk 3 above)
Comment (saneshark) [inline lib/va/api/common/structured_logger/structured_logger.rb]: meant to tag the cases in message_hash

File: spec/va/api/common/structured_logger/structured_logger_spec.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        it 'should output json key value pairs' do
+          expect { logger.info message }.to output(
+            %r{"time":"#{datetime_regex}","type":"INFO","method":"GET","path":"\/things","format":"html","controller":"ThingsController","action":"show","status":200,"duration":15.3,"view":0.36,"db":0.0}
+          ).to_stdout_from_any_process
+        end
+      end
+      context 'with an active record subscriber message' do
+        let(:message) do
+          {
+            sql: 'CREATE TABLE "users" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "first_name" varchar, "last_name" varchar, "age" integer, "created_at" datetime NOT NULL, "updated_at" datetime NOT NULL)',
```
Comment (saneshark) [inline spec/va/api/common/structured_logger/structured_logger_spec.rb]: I'm a little surprised that rubocop is not yelling at you about line length exceeding 120 or whatever it is set at. :)

**Comment by kreek** [comment]
@saneshark thx for the CR, yes most are overridden logger methods, the highlighting is optional if you don't inject it in the json_formatter then it's ignored (I'll update rdoc class comment making this more clear). 👍 on other comments will make fix.

File: spec/va/api/common/structured_logger/structured_logger_spec.rb (hunk 4 above)
Comment (kreek) [inline spec/va/api/common/structured_logger/structured_logger_spec.rb]: I disabled line length for this file, see line #3 😁 wasn't for this line in particular, but for the json regex expectations I can reenable if it's annoying.

Comment (saneshark) [inline spec/va/api/common/structured_logger/structured_logger_spec.rb]: I'm personally cool with it because i use a GUI currently, but maybe someone working in VI or something would have a problem with this.

**Comment by saneshark** [comment]
No problem, one other thing that just occurred to me -- will there still be mechanism for filtering certain entries like password, password_confirmation, social_security, etc? It might not be in the purvue of this gem to determine those types of things, or alternately perhaps it should be?

**Comment by saneshark** [comment]
some additional notes for sake of posterity: 

https://github.com/rails/rails/issues/21404

and for further context on Exception#cause: http://blog.honeybadger.io/nested-errors-in-ruby-with-exception-cause/

**Comment by shauni** [comment]
LGTM :)
